In [15]:
import sys
import os
sys.path.append(os.path.abspath('..'))

# Binary Search

In [16]:
from algorithms.binary_search.inverted_index import InvertedIndex

In [17]:
inverted_index = InvertedIndex()

In [18]:
index = inverted_index.inverted_index_complete_pipeline

### Inverted Index

Decidimos hacer el inverted index con un parametro de ocurrences para saber si es necesario o no (Esto se hizo pensando en acceder de forma mas rapida a los recursos del indice invertido dependienod de si es RRDV o Binary search)

In [19]:
index = inverted_index.inverted_index_complete_pipeline()
index

2024-09-04 15:46:12,987 - InvertedIndex - INFO - Starting complete inverted index pipeline
2024-09-04 15:46:12,987 - InvertedIndex - INFO - Starting complete inverted index pipeline
2024-09-04 15:46:12,987 - INFO - Starting complete inverted index pipeline
2024-09-04 15:46:13,004 - InvertedIndex - INFO - Step 1: Processing texts
2024-09-04 15:46:13,004 - InvertedIndex - INFO - Step 1: Processing texts
2024-09-04 15:46:13,004 - INFO - Step 1: Processing texts
2024-09-04 15:46:13,017 - TextProcessor - INFO - Starting text processing pipeline
2024-09-04 15:46:13,017 - TextProcessor - INFO - Starting text processing pipeline
2024-09-04 15:46:13,017 - TextProcessor - INFO - Starting text processing pipeline
2024-09-04 15:46:13,017 - INFO - Starting text processing pipeline
2024-09-04 15:46:13,041 - TextProcessor - INFO - Reading files from directory: data/docs-raw-texts/
2024-09-04 15:46:13,041 - TextProcessor - INFO - Reading files from directory: data/docs-raw-texts/
2024-09-04 15:46:13,0

{'001': ['218'],
 '01': ['305'],
 '0314': ['098'],
 '044': ['099'],
 '0511': ['125'],
 '06': ['167'],
 '085': ['118'],
 '10': ['012',
  '023',
  '037',
  '038',
  '043',
  '048',
  '052',
  '080',
  '084',
  '091',
  '098',
  '109',
  '112',
  '118',
  '128',
  '130',
  '139',
  '170',
  '199',
  '200',
  '208',
  '209',
  '214',
  '231',
  '234',
  '239',
  '240',
  '242',
  '245',
  '269',
  '293',
  '302',
  '307',
  '309',
  '310',
  '313',
  '324'],
 '100': ['005',
  '070',
  '080',
  '085',
  '090',
  '150',
  '159',
  '167',
  '175',
  '176',
  '177',
  '179',
  '180',
  '194',
  '199',
  '223',
  '230',
  '244',
  '257',
  '268',
  '288'],
 '1000': ['021',
  '023',
  '024',
  '141',
  '145',
  '157',
  '169',
  '180',
  '184',
  '253',
  '269',
  '282'],
 '10000': ['040', '116', '150'],
 '100000': ['115', '116', '204', '240', '258'],
 '1000volt': ['076'],
 '100yen': ['204'],
 '101': ['090', '255'],
 '102': ['145'],
 '1024': ['303'],
 '103': ['098'],
 '1030': ['199'],
 '104': ['

Ahora traemos los datos de los queries y de los documentos ya procesada, esto se hace desde sus respectivas clases

In [20]:
from algorithms.binary_search.query_processor import QueryProcessor
from algorithms.binary_search.text_processor import TextProcessor

texts = TextProcessor()
queries = QueryProcessor()

text_df = texts.process_texts()
queries_df = queries.process_queries()


2024-09-04 15:49:53,111 - INFO - QueryProcessor initialized
2024-09-04 15:49:53,115 - TextProcessor - INFO - Starting text processing pipeline


2024-09-04 15:49:53,115 - TextProcessor - INFO - Starting text processing pipeline
2024-09-04 15:49:53,115 - TextProcessor - INFO - Starting text processing pipeline
2024-09-04 15:49:53,115 - TextProcessor - INFO - Starting text processing pipeline
2024-09-04 15:49:53,115 - INFO - Starting text processing pipeline
2024-09-04 15:49:53,196 - TextProcessor - INFO - Reading files from directory: data/docs-raw-texts/
2024-09-04 15:49:53,196 - TextProcessor - INFO - Reading files from directory: data/docs-raw-texts/
2024-09-04 15:49:53,196 - TextProcessor - INFO - Reading files from directory: data/docs-raw-texts/
2024-09-04 15:49:53,196 - TextProcessor - INFO - Reading files from directory: data/docs-raw-texts/
2024-09-04 15:49:53,196 - INFO - Reading files from directory: data/docs-raw-texts/
2024-09-04 15:49:53,515 - TextProcessor - INFO - Finished reading files. Time taken: 0.30 seconds
2024-09-04 15:49:53,515 - TextProcessor - INFO - Finished reading files. Time taken: 0.30 seconds
2024

En el procesamiento se esta haciendo los pasos de:
* tokenizacion
* lowercase
* remover stop_words
* Stemming

viendose los resultados como

In [21]:
text_df.head(5)

identifier                                               text  \
0        038  evangelista torricelli 16081647 octob 15 1608 ...   
1        004  mickey mous star walk fame imag flickr user fr...   
2        010  sir jame young simpson 1st baronet 18111870 no...   
3        206  neil postman 1931 2003 march 8 1931 media theo...   
4        212  georg bool 18151864 decemb 8 1864 british math...   

                                           text_list  
0  [evangelista, torricelli, 16081647, octob, 15,...  
1  [mickey, mous, star, walk, fame, imag, flickr,...  
2  [sir, jame, young, simpson, 1st, baronet, 1811...  
3  [neil, postman, 1931, 2003, march, 8, 1931, me...  
4  [georg, bool, 18151864, decemb, 8, 1864, briti...

In [22]:
queries_df.head(5)

identifier                                          query  \
0        q38  Religious beliefs of scientists and explorers   
1        q10                early telecommunication methods   
2        q04               University of Edinburgh research   
3        q07                             Walk of Fame stars   
4        q13               famous members of the Royal Navy   

                       processed_query                           query_list  
0  [religi, belief, scientist, explor]  [religi, belief, scientist, explor]  
1          [earli, telecommun, method]          [earli, telecommun, method]  
2       [univers, edinburgh, research]       [univers, edinburgh, research]  
3                   [walk, fame, star]                   [walk, fame, star]  
4        [famous, member, royal, navi]        [famous, member, royal, navi]

## Consultas BOOLEANAS

### AND

Para el AND hicimos uso de los metodos de intersect dentro de binary_search.py donde hayamos, como dice su nombre, la interseccion  entre 2 listas como si fuera una busqueda de AND. Esto se realiza para todas las queries con documento iterando sobre sus terminos en el indice invertido. Este metodo lo podemos ver en funcionamiento desde la clase de search_engine.py donde finalmente hacemos el llamado para procesar los queries y arrojar los resultados

In [ ]:
index

In [23]:
from algorithms.binary_search.search_engine import SearchEngine


query_processor = QueryProcessor()

search_engine = SearchEngine()
output_file = "results/BSII-AND-queries_result.txt"
search_engine.generate_results_file(queries_df, output_file)

with open(output_file, 'r') as f:
    print(f.read())


2024-09-04 15:50:03,315 - INFO - QueryProcessor initialized
2024-09-04 15:50:03,378 - INFO - SearchEngine initialized
2024-09-04 15:50:03,401 - INFO - Results file generated: results/BSII-AND-queries_result.txt


q01
q02 d231
q03 d286
q04 d004,d034
q05
q06 d250,d277
q07 d026,d029,d069,d257,d297,d303,d329
q08 d291,d293
q09 d132,d150,d176,d184,d229,d250,d277
q10 d121,d271
q11 d105,d147,d152,d283,d291,d318
q12
q13 d037,d046,d294
q14 d136,d174
q15
q16 d094,d133
q17 d105
q18 d029,d185
q19
q20 d150,d174
q21
q22 d179
q23 d020,d032,d167,d211
q24 d129,d221,d240,d282
q25 d192,d194,d203,d210
q26
q27 d025,d031,d090,d139,d254
q28
q29
q30 d169
q31 d257,d265
q32
q33
q34 d108,d110,d117,d205,d251
q35 d198,d205,d223



### NOT

Para una comparación NOT, se realizó una consulta booleana, adaptada al algoritmo de mezcla, en la cual se extraía primero todas las páginas en las cuales aparecía la palabra, y luego recorrer todo el índice invertido, si todos los documentos de esa palabra no coinciden con ningún documento de la palabra escogida, entonces se añade la palabra a la respuesta.

In [27]:
from algorithms.binary_search.binary_search import BinarySearch

binary_search = BinarySearch()

In [30]:
index.keys()

dict_keys(['001', '01', '0314', '044', '0511', '06', '085', '10', '100', '1000', '10000', '100000', '1000volt', '100yen', '101', '102', '1024', '103', '1030', '104', '105', '1057', '106', '10cm', '10foot', '10person', '10th', '11', '1100', '1100km', '1103', '11046', '111', '1120', '1122', '113', '114', '114000', '1147', '115', '1152', '1156', '1158', '1159', '116', '1166', '1183', '1187', '1190', '1199', '11th', '11thcenturi', '12', '120', '1200', '12000', '1200014000', '123', '1250m', '1252', '127000', '128', '1282', '129915', '12engin', '12inchwid', '12th', '12thcenturi', '12year', '13', '130', '1300', '13000', '13000000', '130pm', '131', '132', '134', '136', '1360', '1394', '13th', '14', '140', '1400', '140000', '14031437', '1404', '1407', '1420', '1428', '1430', '1437', '1438', '1447', '145', '1450', '1450s', '1451', '1460s', '1476', '1477', '14771547', '1480', '1480s', '1481', '1482', '1493', '1494', '1498', '1499', '14acr', '14th', '14year', '15', '150', '1500', '15000', '150000'

Para comprobar su correcto funcionamiento, vamos a comprobarlo con 3 casos, uno con la palabra 10, otro con la palabra aachen y otro con la palabra academia, para cada palabra revisamos los documentos en los que está incluido y en la salida de esta función verificamos que estos documentos no estén incluidos

Con la palabra academia

In [31]:
index['academia']

['031']

In [32]:
print((binary_search.not_word('academia',index)))

{'academia': ['001', '002', '003', '004', '005', '006', '007', '008', '009', '010', '011', '012', '013', '014', '015', '016', '017', '018', '019', '020', '021', '022', '023', '024', '025', '026', '027', '028', '029', '030', '032', '033', '034', '035', '036', '037', '038', '039', '040', '041', '042', '043', '044', '045', '046', '047', '048', '049', '050', '051', '052', '053', '054', '055', '056', '057', '058', '059', '060', '061', '062', '063', '064', '065', '066', '067', '068', '069', '070', '071', '072', '073', '074', '075', '076', '077', '078', '079', '080', '081', '082', '083', '084', '085', '086', '087', '088', '089', '090', '091', '092', '093', '094', '095', '096', '097', '098', '099', '100', '101', '102', '103', '104', '105', '106', '107', '108', '109', '110', '111', '112', '113', '114', '115', '116', '117', '118', '119', '120', '121', '122', '123', '124', '125', '126', '127', '128', '129', '130', '131', '132', '133', '134', '135', '136', '137', '138', '139', '140', '141', '142',

Con estos comandos, comprobamos que el documento 031, en el cual aparece la palabra academia, no está incluida en toda la respuesta, y los demás documentos si


Con la palabra aachen

In [26]:
index['aachen']

['139', '161', '252']

In [34]:
print((binary_search.not_word('aachen',index)))

{'aachen': ['001', '002', '003', '004', '005', '006', '007', '008', '009', '010', '011', '012', '013', '014', '015', '016', '017', '018', '019', '020', '021', '022', '023', '024', '025', '026', '027', '028', '029', '030', '031', '032', '033', '034', '035', '036', '037', '038', '039', '040', '041', '042', '043', '044', '045', '046', '047', '048', '049', '050', '051', '052', '053', '054', '055', '056', '057', '058', '059', '060', '061', '062', '063', '064', '065', '066', '067', '068', '069', '070', '071', '072', '073', '074', '075', '076', '077', '078', '079', '080', '081', '082', '083', '084', '085', '086', '087', '088', '089', '090', '091', '092', '093', '094', '095', '096', '097', '098', '099', '100', '101', '102', '103', '104', '105', '106', '107', '108', '109', '110', '111', '112', '113', '114', '115', '116', '117', '118', '119', '120', '121', '122', '123', '124', '125', '126', '127', '128', '129', '130', '131', '132', '133', '134', '135', '136', '137', '138', '140', '141', '142', '

Comprobamos que los documentos 039, 161 y 252 no aparecen en la respuesta, y los demás documentos si


Con la palabra 10

In [35]:
index['10']

['012',
 '023',
 '037',
 '038',
 '043',
 '048',
 '052',
 '080',
 '084',
 '091',
 '098',
 '109',
 '112',
 '118',
 '128',
 '130',
 '139',
 '170',
 '199',
 '200',
 '208',
 '209',
 '214',
 '231',
 '234',
 '239',
 '240',
 '242',
 '245',
 '269',
 '293',
 '302',
 '307',
 '309',
 '310',
 '313',
 '324']

In [36]:
print((binary_search.not_word('10',index)))

{'10': ['001', '002', '003', '004', '005', '006', '007', '008', '009', '010', '011', '013', '014', '015', '016', '017', '018', '019', '020', '021', '022', '024', '025', '026', '027', '028', '029', '030', '031', '032', '033', '034', '035', '036', '039', '040', '041', '042', '044', '045', '046', '047', '049', '050', '051', '053', '054', '055', '056', '057', '058', '059', '060', '061', '062', '063', '064', '065', '066', '067', '068', '069', '070', '071', '072', '073', '074', '075', '076', '077', '078', '079', '081', '082', '083', '085', '086', '087', '088', '089', '090', '092', '093', '094', '095', '096', '097', '099', '100', '101', '102', '103', '104', '105', '106', '107', '108', '110', '111', '113', '114', '115', '116', '117', '119', '120', '121', '122', '123', '124', '125', '126', '127', '129', '131', '132', '133', '134', '135', '136', '137', '138', '140', '141', '142', '143', '144', '145', '146', '147', '148', '149', '150', '151', '152', '153', '154', '155', '156', '157', '158', '159'

Vemos que los documentos en los cuales aparece la palabra '10' no están incluidos en la respuesta, entre ellos 012, 023, 027, etc....